server/src/main/scala/org/http4s/server/middleware/Throttle.scala

```diff
@@ -0,0 +1,90 @@
+package org.http4s.server.middleware
+
+import org.http4s.{Http, Response, Status}
+import cats.data.Kleisli
+import cats.effect.{Concurrent, Timer}
+import cats.effect.concurrent.Ref
+import scala.concurrent.duration._
+import cats.implicits._
+import fs2.Stream
+
+sealed trait TokenAvailability
```

Comment:
`sealed abstract class TokenAvailability extends Product with Serializable`: noisier, but better inference an bin compat.

This is dumping a bunch of symbols into the middleware package.  I wonder whether this ADT, and the `TokenBucket`, should be moved inside `Throttle`.  Or maybe, in the opposite direction, it could be useful on a `Client`, and we should move it into `org.http4s.util`.  :thinking:

Reply:
Slight preference for having it into `Throttle` and not `util` for the time being.

Reply:
But move `Throttle` to another package or leave it here?  I agree with moving the ADT into `Throttle` regardless.

`org.http4s.utill` is a cursed place of "things that make us wish that Scala were more amenable to microlibraries," so I do acknowledge the disgust.

Reply:
I won't oppose moving into `util` once we have the need. For now I'd rather keep it here so that the code is easier to follow. 

On a related not wrt reusability, client-side limiting imho is not a great fit for a token bucket. `upperbound` (which focuses on that) has a different design for that reason